### ISSUE TITLE
HTTP/2 HPACK header table size update not understood by the client

### ISSUE THREAD

**Issue by trendspotter** [openMF/android-client#1005]
**Summary:** 

MifosXdroid or one of the dependencies does not understand specific HTTP/2 protocol feature.

**Steps to reproduce:** 

1. Install MifosX on Apache Tomcat
2. Set nginx version 1.13.6 or newer as reverse proxy for the Tomcat
3. Enable HTTP/2 and TLS on the nginx reverse proxy
4. Attempt to connect to the instance using MifosXdroid client

**Expected behaviour:** 

Client can connect (ie. either successfully log in or display a message about wrong credentials)

**Observed behaviour:** 

Clien't can't connect to the instance. It receives data, but nothing is shown to the user and application remains on login screen.

This may be a bit low-level, so bear with me... MifosXdroid v1.0.1 does not understand properly a web server response containing **HTTP/2 HPACK header table size update** information. We used to have Mifos X behind a nginx **1.12.2** reverse proxy with TLS and HTTP/2 enabled. After we have upgraded nginx to **1.14.1**, MifosXdroid was no longer able to connect to the server. No error message was displayed on either client or server side. In fact no messages were displayed at all, as if the response was discarded by the client. Server logs indicated that the requests and responses are accepted and returned as expected.

Finally, I have traced the connections using tcpdump and wireshark and decrypted them to see the differences between the old and the new setup. The only meaningful difference was, that nginx since version 1.13.6 (commit https://github.com/nginx/nginx/commit/63f5d46f58e537745987e9d68bc65b03830238a7) adds HTTP/2 0-byte HPACK table size update information to inform that it doesn't use HPACK's dynamic table (which is in accordance with [RFC7540](https://tools.ietf.org/html/rfc7540) section 6.5.2 and [RFC 7541](https://tools.ietf.org/html/rfc7541)). When I stripped this information using intercepting transparent proxy (mitmproxy), the client was able to connect again.

This is presumably not a problem of the codebase but of one of the third-party dependencies (most likely *okhttp*). In order to fix the problem, this dependency either needs to be updated or, if the bug is present even in the newest version, reported in the appropriate upstream. (I have zero experience with Android development to do so myself, I'm a sysadmin)

**Device and Android version:** 

HTC 10, stock ROM, Android 8.0.0. Tested also on bunch of others.

**Screenshots:** 

![hpack](https://user-images.githubusercontent.com/15286128/49697027-a6716580-fbb2-11e8-80ac-a1177697b40e.png)

**Comment by trendspotter** [comment]
After some more poking around, I'm 99% sure that this is caused by the outdated OkHttp component. MifosXdroid 1.0.1 uses okhttp/**3.3.1**. According to the [changelog](https://github.com/square/okhttp/blob/master/CHANGELOG.md#version-340), HPACK dynamic table size is supported from **3.4.0**, but that version had off-by-one error, so full support for HPACK is there since **3.5.0**.

OkHttp needs to be updated.

**Comment by trendspotter** [comment]
Thanks to your wiki it wasn't so hard to get it compiled and tested. I've checked okHttp 3.4.2, which still failed and then okHttp 3.5.0 which finally worked correctly even with the HPACK table size update shown above. I figured that even though there is already much newer version of okHttp, the rest of the MifosXdroid codebase is still old, so 3.5.0 seemed like the path of the least resistance. Pushed a PR #1006.
